fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/identifiers/FedoraID.java

```diff
@@ -156,6 +180,56 @@ public String getMementoString() {
         return mementoDatetimeStr;
     }
 
+    /**
+     * Add a part onto this identifier.
+     * @param addition The additional string.
+     * @return new FedoraID for the new identifier.
+     */
+    public FedoraID addToResourceId(final String addition) {
+        return FedoraID.create(this.getResourceId() + (this.getResourceId().endsWith("/") ? "" : "/") + addition);
+    }
+
+    /**
+     * Add a part onto this identifier.
+     * @param addition The additional string.
+     * @return new FedoraID for the new identifier.
+     */
+    public FedoraID addToFullId(final String addition) {
+        return FedoraID.create(this.getFullId() + (this.getFullId().endsWith("/") ? "" : "/") + addition);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (obj == this) {
+            return true;
+        }
+
+        if (!(obj instanceof FedoraID)) {
+            return false;
+        }
+
+        final var testObj = (FedoraID) obj;
+        return testObj.getFullId().equals(this.getFullId());
+    }
+
+    @Override
+    public int hashCode() {
+        return (int) getFullId().hashCode();
+    }
+
+    /**
+     * Ensure the ID has the info:fedora/ prefix.
+     * @param id the identifier, if null assume repository root (info:fedora/)
+     * @return the identifier with the info:fedora/ prefix.
+     */
+    private static String ensurePrefix(final String id) {
+        if (id == null) {
+            return FEDORA_ID_PREFIX;
+        }
+        final String decodedId = URLDecoder.decode(id, UTF_8);
```

Comment:
Is this safe? I would have expect a resource id to have been decoded somewhere in the HTTP layer if it needed to be.

Reply:
I can move it to the HttpIdentifierConverter if you prefer

Reply:
I think that might be better. I have a fear of decoding strings that shouldn't be decoded.